pkg/util/liveconfig/dev.go

```diff
@@ -12,24 +12,45 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 )
 
-func (d *dev) HiveRestConfig(ctx context.Context, index int) (*rest.Config, error) {
-	// Indexes above 0 have _index appended to them
+func (d *dev) HiveRestConfig(ctx context.Context, shard int) (*rest.Config, error) {
+	// shards above 0 have _shard appended to it
 	envVar := hiveKubeconfigPathEnvVar
-	if index != 0 {
-		envVar = fmt.Sprintf("%s_%d", hiveKubeconfigPathEnvVar, index)
+	if shard != 0 {
+		envVar = fmt.Sprintf("%s_%d", hiveKubeconfigPathEnvVar, shard)
 	}
 
+	// Use an override kubeconfig path if one is provided
 	kubeConfigPath := os.Getenv(envVar)
-	if kubeConfigPath == "" {
-		return nil, fmt.Errorf("missing %s env variable", hiveKubeconfigPathEnvVar)
+	if kubeConfigPath != "" {
+		restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+		if err != nil {
+			return nil, err
+		}
+		return restConfig, nil
 	}
 
-	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	// Hive shards are planned but not deployed or implemented yet
+	d.hiveCredentialsMutex.RLock()
+	cached, exists := d.cachedCredentials[shard]
+	d.hiveCredentialsMutex.RUnlock()
+	if exists {
+		return rest.CopyConfig(cached), nil
+	}
+
+	// Lock the RWMutex as we're starting to fetch so that new readers will wait
+	// for the existing Azure API call to be done.
+	d.hiveCredentialsMutex.Lock()
+
```

Comment:
nit: `hive.go` uses defer for unlocking. Can we update `dev.go` to match?

Reply:
Missed that, thanks.